### PR TITLE
First draft of attention

### PR DESCRIPTION
**Pull request by derekchen14**
Able to run train and test, although results don't look great yet.  Need to tune params and get the neural session to work properly.

### CODE REVIEW

File: negotiation/Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 model=encdec
-exp=basic
-lr=0.5
+exp=decay
+lr=0.031
```
Comment (hhexiy) [inline negotiation/Makefile]: you don't have to change the default in the file. you can specify these in cmd line, e.g. `make train lr=0.1`. let's keep the default value in the file.

File: negotiation/neural/attention.py [hunk 2]
```diff
@@ -193,3 +193,99 @@ def forward(self, input, memory_bank, memory_lengths=None, coverage=None):
...
+        else:
+            one_step = False
+
+        attention_hidden_states = []
+        alignment_vectors = []
+
+        for idx, memory_bank in enumerate(memory_banks):
+            batch, sourceL, dim = memory_bank.size()
+            batch_, targetL, dim_ = input.size()
+            aeq(batch, batch_)
```
Comment (hhexiy) [inline negotiation/neural/attention.py]: bad var name. don't use camel case. be consistent with onmt.

File: negotiation/neural/attention.py [hunk 3]
```diff
@@ -193,3 +193,99 @@ def forward(self, input, memory_bank, memory_lengths=None, coverage=None):
...
+            aeq(dim, dim_)
+            aeq(self.dim, dim)
+            if coverage is not None:
```
Comment (hhexiy) [inline negotiation/neural/attention.py]: why can't you just call GlobalAttention instead of copying over the code?

File: negotiation/neural/generator.py [hunk 4]
```diff
@@ -113,7 +124,10 @@ def unbottle(m):
         # (2) Repeat src objects `beam_size` times.
         #src_map = rvar(batch.src_map.data) \
         #    if data_type == 'text' and self.copy_attn else None
-        memory_bank = rvar(memory_bank.data)
+        if predict_with_context:
+            memory_bank = [rvar(mem_bank.data).transpose(0,1) for mem_bank in memory_bank]
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: why do we need transpose here?

File: negotiation/neural/generator.py [hunk 5]
```diff
@@ -102,6 +102,17 @@ def unbottle(m):
...
+        if hasattr(batch, 'prev_turns'):
+            item_title = batch.item_title
+            previous_turns = batch.prev_turns
+
+            _, item_memory_bank = self.model.cbow_embedder(item_title)
+            _, prev_memory_bank = self.model.cbow_embedder(previous_turns)
+            enc_memory_bank = memory_bank
+
+            memory_bank = [enc_memory_bank, prev_memory_bank.transpose(0,1)]
+            predict_with_context = True
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: don't do adhoc option here. use model option to decide if we're using context, e.g. check num_context

File: negotiation/neural/model_builder.py [hunk 6]
```diff
@@ -31,8 +31,8 @@ def add_model_arguments(parser):
                        help="""Type of encoder layer to use. Non-RNN layers
                        are experimental. Options are
                        [rnn|brnn|mean|transformer|cnn].""")
-    group.add_argument('--decoder-type', type=str, default='rnn',
-                       choices=['rnn', 'transformer', 'cnn'],
+    group.add_argument('--decoder-type', type=str, default='multibank',
+                       choices=['rnn', 'transformer', 'cnn', 'multibank'],
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: the multibank should be used to specify attention type. the decoder is still rnn. can you check opt.global_attention? can we specify different types of attention there?

File: negotiation/neural/model_builder.py [hunk 7]
```diff
@@ -141,13 +141,18 @@ def make_decoder(opt, embeddings):
...
                              attn_type=opt.global_attention,
                              dropout=opt.dropout,
                              embeddings=embeddings)
-
+    else:
+        return MultiAttnDecoder(opt.rnn_type, bidirectional,
+                             opt.dec_layers, opt.rnn_size,
+                             attn_type=opt.global_attention,
+                             dropout=opt.dropout,
+                             embeddings=embeddings)
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py:168]: see above comment. specify attn_type, don't add a new type of decoder.

File: negotiation/neural/model_builder.py [hunk 8]
```diff
@@ -180,26 +185,23 @@ def make_base_model(model_opt, mappings, gpu, checkpoint=None):
...
-
     encoder = make_encoder(model_opt, src_embeddings)
-
+    # Make context embedder.
+    original_enc_type = model_opt.encoder_type
+    model_opt.encoder_type = "mean"
+    context_dict = mappings['vocab']
+    context_embeddings = make_embeddings(model_opt, context_dict)
+    context_embedder = make_encoder(model_opt, context_embeddings)
+    model_opt.encoder_type = original_enc_type
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: should check if we are using the context before making context embedder.

File: onmt/Optim.py [hunk 9]
```diff
@@ -58,7 +56,8 @@ def set_parameters(self, params):
         if self.method == 'sgd':
             self.optimizer = optim.SGD(self.params, lr=self.lr)
         elif self.method == 'adagrad':
-            self.optimizer = optim.Adagrad(self.params, lr=self.lr)
+            self.optimizer = optim.Adagrad(self.params, lr=self.lr,
+                weight_decay=self.weight_decay)
```
Comment (hhexiy) [inline onmt/Optim.py]: why do you need weight_decay for adagrad?

File: negotiation/neural/models.py [hunk 10]
```diff
@@ -282,13 +282,13 @@ def __init__(self, rnn_type, bidirectional_encoder, num_layers,
...
-    def forward(self, tgt, memory_bank, state, memory_lengths=None):
+    def forward(self, tgt, memory_banks, state, memory_lengths=None):
         """
         Args:
             tgt (`LongTensor`): sequences of padded tokens
                                 `[tgt_len x batch]`.
-            memory_bank (`FloatTensor`): vectors from the encoder
-                 `[src_len x batch x hidden]`.
+            memory_bank(s) (`FloatTensor`): vectors from the encoder
+                 `[src_len x batch x hidden]`. possibly a list of vectors
```
Comment (hhexiy) [inline negotiation/neural/models.py]: why did you change the dimension? it's supposed to be time major.

File: negotiation/neural/models.py [hunk 11]
```diff
@@ -93,8 +94,7 @@ def forward(self, src, lengths=None, encoder_state=None):
         encoder_final = (mean, mean)
         return encoder_final, memory_bank
 
-
-class RNNEncoder(EncoderBase):
+class StdRNNEncoder(EncoderBase):
```
Comment (hhexiy) [inline negotiation/neural/models.py:99]: let's not change their original class but create an inheritance.

File: negotiation/neural/models.py [hunk 12]
```diff
@@ -304,13 +304,21 @@ def forward(self, tgt, memory_bank, state, memory_lengths=None):
...
         tgt_len, tgt_batch = tgt.size()
-        _, memory_batch, _ = memory_bank.size()
+        if isinstance(memory_banks, list):
+            memory_bank = memory_banks[0]
+        else:
+            # We transpose since the code originally did so when calculating
+            # attention in StdRNNDecoder._run_forward_pass(), but we can no
+            # longer do that anymore because memory_banks is sometimes a list
+            memory_banks = memory_banks.transpose(0, 1)
+            memory_bank = memory_banks
```
Comment (hhexiy) [inline negotiation/neural/models.py]: this part is a bit sketchy. they use time major (seq_len as first dimension) for all inputs. let's not change that.

File: negotiation/neural/models.py [hunk 13]
```diff
@@ -596,6 +603,26 @@ def forward(self, src, tgt, lengths, dec_state=None):
             attns = None
         return decoder_outputs, attns, dec_state
 
+class NegotiationModel(NMTModel):
+
+    def __init__(self, encoder, decoder, cbow_embedder):
```
Comment (hhexiy) [inline negotiation/neural/models.py]: cbow_embedder -> context_embedder. it doesn't have to be cbow

File: negotiation/neural/models.py [hunk 14]
```diff
@@ -596,6 +603,26 @@ def forward(self, src, tgt, lengths, dec_state=None):
...
+        super(NegotiationModel, self).__init__(encoder, decoder)
+        self.cbow_embedder = cbow_embedder
+
+    def forward(self, sources, tgt, lengths, dec_state=None):
+        enc_final, enc_memory_bank = self.encoder(sources['enc'], lengths)
+        # item_final, item_memory_bank = self.cbow_embedder(sources['item'])
+        prev_final, prev_memory_bank = self.cbow_embedder(sources['prev'])
```
Comment (hhexiy) [inline negotiation/neural/models.py]: if we don't use prev_final, make it _.

File: negotiation/neural/models.py [hunk 15]
```diff
@@ -596,6 +603,26 @@ def forward(self, src, tgt, lengths, dec_state=None):
...
+        # memory_banks = [enc_memory_bank, item_memory_bank, prev_memory_bank]
+        memory_banks = [enc_memory_bank.transpose(0,1), prev_memory_bank]
```
Comment (hhexiy) [inline negotiation/neural/models.py]: weird to have transpose here. do transpose at one place and write the tensor size in comment, e.g. (batch, context_len, dim)

File: negotiation/neural/trainer.py [hunk 16]
```diff
@@ -235,7 +235,13 @@ def validate(self, valid_iter):
...
             lengths = batch.lengths
 
-            outputs, attns, _ = self.model(encoder_inputs, decoder_inputs, lengths)
+            item_title = batch.item_title
+            previous_turns = batch.prev_turns
+            combined_inputs = {'enc': encoder_inputs,
+                              'item': item_title,
+                              'prev': previous_turns}
+
+            outputs, attns, _ = self.model(combined_inputs, decoder_inputs, lengths)
```
Comment (hhexiy) [inline negotiation/neural/trainer.py]: combined inputs:
separate encoder inputs from context inputs, so you can have encoder_inputs, decoder_inputs, context_inputs, length

File: negotiation/neural/models.py [hunk 17]
```diff
@@ -366,7 +374,7 @@ def _run_forward_pass(self, tgt, memory_bank, state, memory_lengths=None):
             tgt (LongTensor): a sequence of input tokens tensors
                                  [len x batch].
             memory_bank (FloatTensor): output(tensor sequence) from the encoder
-                        RNN of size (src_len x batch x hidden_size).
+                        RNN of size (batch x src_len x hidden_size).
```
Comment (hhexiy) [inline negotiation/neural/models.py]: can you check the dim here

File: onmt/Optim.py [hunk 18]
```diff
@@ -31,13 +31,9 @@ class Optim(object):
...
-    def __init__(self, method, lr, max_grad_norm,
-                 lr_decay=1, start_decay_at=None,
-                 beta1=0.9, beta2=0.999,
-                 adagrad_accum=0.0,
-                 decay_method=None,
-                 warmup_steps=4000,
-                 model_size=None):
+    def __init__(self, method, lr, max_grad_norm, lr_decay=0.9, start_decay_at=None,
+                 beta1=0.9, beta2=0.999, adagrad_accum=0.0,
+                 decay_method=None, warmup_steps=4000, model_size=None):
```
Comment (hhexiy) [inline onmt/Optim.py]: is there a reason to change its default setting?

File: negotiation/sessions/neural_session.py [hunk 19]
```diff
@@ -269,6 +271,14 @@ def _is_valid(self, tokens):
...
+        clean_tokens = []
+        for pred in predictions:
+            token = self.vocab.ind_to_word[pred]
+            if is_entity(token):
+                token = self.dialogue.entity_to_price(self.kb, token)
+            clean_tokens.append(token)
+            if clean_tokens[-1] == markers.EOS:
+                clean_tokens = clean_tokens[:-1]
+                break
+        return clean_tokens
```
Comment (hhexiy) [inline negotiation/sessions/neural_session.py]: can we put this in _build_target_token?

File: negotiation/neural/generator.py [hunk 20]
```diff
@@ -102,7 +102,17 @@ def unbottle(m):
         enc_states, memory_bank = self.model.encoder(encoder_inputs, lengths)
         dec_states = self.model.decoder.init_decoder_state(
                                         encoder_inputs, memory_bank, enc_states)
-        # enc/dec_states: (seq_len, batch_size, rnn_size)
+
+        if batch.num_context > 1:
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: this should be > 0

File: negotiation/neural/generator.py [hunk 21]
```diff
@@ -113,7 +123,11 @@ def unbottle(m):
         # (2) Repeat src objects `beam_size` times.
         #src_map = rvar(batch.src_map.data) \
         #    if data_type == 'text' and self.copy_attn else None
-        memory_bank = rvar(memory_bank.data)
+        if batch.num_context > 1:
+            memory_bank[0] = rvar(memory_bank[0].data)
+            memory_bank[1] = rvar(memory_bank[1].transpose(0,1).data).transpose(0,1)
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: why do you do 2 transpose here

Comment (hhexiy) [reply negotiation/neural/generator.py]: also can you add a comment about what's in the list memory_bank and the dim of each element

File: negotiation/neural/batcher.py [hunk 22]
```diff
@@ -28,8 +31,9 @@ def __init__(self, encoder_args, decoder_args, context_data, vocab,
...
             for k, v in self.context_data.iteritems():
                 self.context_data[k] = self.order_by_id(v, sorted_ids)
-            for attr in ('encoder_inputs', 'decoder_inputs', 'targets', 'lengths'):
-                setattr(self, attr, self.order_by_id(getattr(self, attr), sorted_ids))
+            for attr in ('encoder_inputs', 'decoder_inputs', 'item_inputs', 'context_inputs', 'targets', 'lengths'):
+                sorted_attrs = self.order_by_id(getattr(self, attr), sorted_ids)
+                setattr(self, attr, sorted_attrs)
 
         if time_major:
             for attr in ('encoder_inputs', 'decoder_inputs', 'targets'):
```
Comment (hhexiy) [inline negotiation/neural/batcher.py]: make context inputs time_major too. so it is processed in the same way as encoder inputs later in the code.

File: negotiation/neural/models.py [hunk 23]
```diff
@@ -399,11 +403,22 @@ def _run_forward_pass(self, tgt, memory_bank, state, memory_lengths=None):
...
         decoder_outputs, p_attn = self.attn(
-            rnn_output.transpose(0, 1).contiguous(),
-            memory_bank.transpose(0, 1),
+            rnn_output.transpose(0, 1).contiguous(), memory_banks,
             memory_lengths=memory_lengths
         )
+        if isinstance(memory_banks, list):
+            # change the dimensions back so generator can run
+            memory_banks[0] = memory_banks[0].transpose(0,1)
+
```
Comment (hhexiy) [inline negotiation/neural/models.py]: is this part redundant?

File: negotiation/neural/generator.py (hunk 21 above)
Comment (hhexiy) [reply negotiation/neural/generator.py]: see comment in Batch.

File: negotiation/neural/model_builder.py [hunk 24]
```diff
@@ -180,26 +183,26 @@ def make_base_model(model_opt, mappings, gpu, checkpoint=None):
     # Make encoder.
     src_dict = mappings['vocab']
     src_embeddings = make_embeddings(model_opt, src_dict)
-
     encoder = make_encoder(model_opt, src_embeddings)
-
+    # Make context embedder.
+    if model_opt.num_context > 1:
+      original_enc_type = model_opt.encoder_type
+      model_opt.encoder_type = "mean"
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: this is a little weird... doesn't changing encoder_type affect later make_embedding? ideally we should have a different arg like "context encoder type" or something

File: onmt/Optim.py (hunk 18 above)
Comment (derekchen14) [reply onmt/Optim.py]: I was experimenting to see if this caused the drop in performance, and now we know it wasn't, so I've changed it back to the original settings.

File: negotiation/sessions/neural_session.py (hunk 19 above)
Comment (derekchen14) [reply negotiation/sessions/neural_session.py]: I tried calling self.builder(pred) at one point, but needed too many hacks to make it work, mostly because the function needs access to self.dialogue and self.kb which utterance builder doesn't hase.  Also, the function needs to extract the predictions from the data which is a different step.

**Comment by derekchen14** [comment]
Resolved the transpose issues and the other items above.  it's fairly clean now with only one "tranpose"  in the StdRNNDecoder.   Let me know if this looks good!

File: negotiation/sessions/neural_session.py (hunk 19 above)
Comment (hhexiy) [reply negotiation/sessions/neural_session.py]: can we add `kb` as a param with default value `None` in `_build_target_token`, so that here you can call it with the additional input `kb`?

Comment (derekchen14) [reply negotiation/sessions/neural_session.py]: I updated session so that it calls on _build_target_tokens.  I feel like the set-up is a bit messy, but maybe there's a cleaner way to write this.   Let me know what you prefer.

File: negotiation/neural/preprocess.py [hunk 25]
```diff
@@ -321,11 +321,14 @@ def to_float_price(entity):
...
-        human_readable_price = "${}".format(raw_price.canonical.value)
-        return human_readable_price
+    def get_ent_to_price(self, kb):
+
+        def entity_to_price(entity_token):
+            raw_price = PriceScaler.unscale_price(kb, entity_token)
+            human_readable_price = "${}".format(raw_price.canonical.value)
+            return human_readable_price
+
+        return entity_to_price
```
Comment (hhexiy) [inline negotiation/neural/preprocess.py]: hmm.. this should be in utterance builder too. also let's keep the original function in preprocess (some other sessions might be still using it). and add this to utterance builder.

File: negotiation/neural/model_builder.py [hunk 26]
```diff
@@ -180,26 +197,21 @@ def make_base_model(model_opt, mappings, gpu, checkpoint=None):
...
     src_embeddings = make_embeddings(model_opt, src_dict)
-
     encoder = make_encoder(model_opt, src_embeddings)
-
+    # Make context embedder.
+    if model_opt.num_context > 1:
+      context_dict = mappings['vocab']
+      context_embeddings = make_embeddings(model_opt, context_dict)
+      context_embedder = make_context_embedder(model_opt, context_embeddings)
     # Make decoder.
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: num_context should be > 0

File: negotiation/sessions/neural_session.py [hunk 27]
```diff
@@ -129,6 +124,7 @@ def _create_batch(self):
                 'encoder_args': encoder_args,
                 'decoder_args': decoder_args,
                 }
+        pdb.set_trace()
```
Comment (hhexiy) [inline negotiation/sessions/neural_session.py]: cleanup debug statements?

File: negotiation/neural/model_builder.py [hunk 28]
```diff
@@ -226,12 +242,26 @@ def make_base_model(model_opt, mappings, gpu, checkpoint=None):
...
-                    model_opt.pretrained_wordvec, model_opt.fix_pretrained_wordvec)
+            if model.encoder.embed_type == 'dialogue':
+              model.encoder.embeddings.load_pretrained_vectors(
+                    dialogue_wordvec, model_opt.fix_pretrained_wordvec)
+            elif model.encoder.embed_type == 'kb':
+              model.encoder.embeddings.load_pretrained_vectors(
+                    kb_wordvec, model_opt.fix_pretrained_wordvec)
+            elif model.encoder.embed_type == 'category':
+              model.encoder.embeddings.load_pretrained_vectors(
+                    cat_wordvec, model_opt.fix_pretrained_wordvec)
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: we don't have category vocab

Comment (hhexiy) [reply negotiation/neural/model_builder.py]: also encoder only uses dialogue_vocab right?

Comment (hhexiy) [reply negotiation/neural/model_builder.py]: we need to load for context_embedder and kb_embedder

File: negotiation/sessions/neural_session.py (hunk 27 above)
Comment (derekchen14) [reply negotiation/sessions/neural_session.py]: done, that seems to be the only one left

File: negotiation/neural/model_builder.py (hunk 28 above)
Comment (derekchen14) [reply negotiation/neural/model_builder.py]: removed the "category" vocab, i'm seeing that your code style is to not write out snippets until we need them.
The encoder and context_embedder both use "dialogue_wordvec", and only "dialogue_wordvec".
The kb_embedder uses "kb_wordvec" and only "kb_wordvec".

The way it works now is each of the 3 embedding models have a "embed_type" attribute signifying the (one) vocab type to load into self.embeddings.  that embed_type is either "dialogue" or "kb".  Thus, we are already loading the pytorch embeddings for all three.

File: negotiation/neural/model_builder.py [hunk 29]
```diff
@@ -256,9 +256,6 @@ def make_base_model(model_opt, mappings, gpu, checkpoint=None):
             elif model.encoder.embed_type == 'kb':
               model.encoder.embeddings.load_pretrained_vectors(
                     kb_wordvec, model_opt.fix_pretrained_wordvec)
-            elif model.encoder.embed_type == 'category':
-              model.encoder.embeddings.load_pretrained_vectors(
-                    cat_wordvec, model_opt.fix_pretrained_wordvec)
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: the embeddings are all loaded to the encoder?? shouldn't it be the embedders?

File: negotiation/evaluate.py [hunk 30]
```diff
@@ -53,8 +53,13 @@
...
-    evaluator = Evaluator(model, mappings['vocab'], gt_prefix=1)
+    
+    # if the pretrained words are a list, then it means there are two vocabs
+    if isinstance(model_args.pretrained_wordvec, list):
+        # so we pass in the entire mappings, which include 'kb_vocab'
+        evaluator = Evaluator(model, mappings, gt_prefix=1)
+    else:
+        # Prefix: [GO, CATEGORY]
+        # Just giving it GO seems okay as it can learn to copy the CATEGORY from the input
+        evaluator = Evaluator(model, mappings['vocab'], gt_prefix=1)
```
Comment (hhexiy) [inline negotiation/evaluate.py]: you can just pass in mappings.

File: negotiation/neural/evaluator.py [hunk 31]
```diff
@@ -39,15 +39,15 @@ def add_evaluator_arguments(parser):
...
+    def __init__(self, model, mappings, gt_prefix=1):
         self.model = model
         self.gt_prefix = gt_prefix
 
         if isinstance(mappings, dict):
             self.vocab = mappings['vocab']
             self.kb_vocab = mappings['kb_vocab']
         else:
-            self.vocab = vocab
+            self.vocab = mappings
```
Comment (hhexiy) [inline negotiation/neural/evaluator.py]: this is redundant. just make mappings a dict.
